api/datastore/redis/redis.go

```diff
@@ -0,0 +1,365 @@
+	if name == "" {
+		return nil, models.ErrDatastoreEmptyAppName
+	}
+
+	reply, err := ds.conn.Do("HGET", "apps", name)
+	if err != nil {
+		return nil, err
+	} else if reply == nil {
+		return nil, models.ErrAppsNotFound
+	}
+
+	res := &models.App{}
+	if err := json.Unmarshal(reply.([]byte), res); err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (ds *RedisDataStore) GetApps(ctx context.Context, filter *models.AppFilter) ([]*models.App, error) {
+	res := []*models.App{}
+
+	reply, err := ds.conn.Do("HGETALL", "apps")
+	if err != nil {
+		return nil, err
+	}
+
+	apps, err := redis.StringMap(reply, err)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range apps {
+		var app models.App
+		if err := json.Unmarshal([]byte(v), &app); err != nil {
+			return nil, err
+		}
+		if applyAppFilter(&app, filter) {
+			res = append(res, &app)
+		}
+	}
+	return res, nil
+}
+
+func (ds *RedisDataStore) setRoute(set string, route *models.Route) (*models.Route, error) {
+	buf, err := json.Marshal(route)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := ds.conn.Do("HSET", set, route.Path, buf); err != nil {
+		return nil, err
+	}
+
+	if _, err := ds.conn.Do("SADD", "routes", set); err != nil {
+		return nil, err
+	}
+
+	return route, nil
+}
+
+func (ds *RedisDataStore) InsertRoute(ctx context.Context, route *models.Route) (*models.Route, error) {
+	if route == nil {
+		return nil, models.ErrDatastoreEmptyRoute
+	}
+
+	if route.AppName == "" {
+		return nil, models.ErrDatastoreEmptyAppName
+	}
+
+	if route.Path == "" {
+		return nil, models.ErrDatastoreEmptyRoutePath
+	}
+	hset := fmt.Sprintf("routes:%s", route.AppName)
+
+	if reply, err := ds.conn.Do("HGET", hset, route.Path); err != nil {
+		return nil, err
+	} else if reply != nil {
+		return nil, models.ErrRoutesAlreadyExists
+	}
+
+	return ds.setRoute(hset, route)
+}
+
+func (ds *RedisDataStore) UpdateRoute(ctx context.Context, route *models.Route) (*models.Route, error) {
+	if route == nil {
+		return nil, models.ErrDatastoreEmptyRoute
+	}
+
+	if route.AppName == "" {
+		return nil, models.ErrDatastoreEmptyAppName
+	}
+
+	if route.Path == "" {
+		return nil, models.ErrDatastoreEmptyRoutePath
+	}
+
+	hset := fmt.Sprintf("routes:%s", route.AppName)
+	if reply, err := ds.conn.Do("HGET", hset, route.Path); err != nil {
+		return nil, err
+	} else if reply == nil {
+		return nil, models.ErrRoutesNotFound
+	}
+
+	return ds.setRoute(hset, route)
+}
+
+func (ds *RedisDataStore) RemoveRoute(ctx context.Context, appName, routePath string) error {
+	if appName == "" {
+		return models.ErrDatastoreEmptyAppName
+	}
+
+	if routePath == "" {
+		return models.ErrDatastoreEmptyRoutePath
+	}
+
+	hset := fmt.Sprintf("routes:%s", appName)
+	if n, err := ds.conn.Do("HDEL", hset, routePath); err != nil {
+		return err
+	} else if n == 0 {
+		return models.ErrRoutesRemoving
+	}
+
+	if _, err := ds.conn.Do("SREM", "routes", hset); err != nil {
```

Comment:
since Redis is a dictionary based DB you need to keep track of all the existing "sets" so you can iterate through them later... check ```GetRoutes```